add customised cart & checkout block settings to wc-tracker snapshot

haszari opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds extra information to the WC Tracker snapshot about the settings used for cart & checkout blocks. If the cart or checkout blocks are used on the relevant pages, any customised attributes are saved to the tracker snapshot. 

For stores that are opted in to tracker, this will allow us to get usage information for how the cart & checkout blocks are being used.

This is a follow up to https://github.com/woocommerce/woocommerce/pull/25932

### How to test the changes in this Pull Request:
1. Install and activate [WooCommerce Blocks plugin from source](https://github.com/woocommerce/woocommerce-gutenberg-products-block) - the cart & checkout block have not been released yet.
2. Add the cart and checkout block to cart & checkout pages and change some settings, e.g. `Show a "return to cart" link`. Ensure those pages are published and are the designated cart & checkout pages (`WooCommerce > Settings > Advanced > Page setup`).
2. Ensure your test site is opted into tracking (`WooCommerce > Settings > Advanced > WooCommerce.com > Usage Tracking`). 

Using `wp shell`, you can call the new method directly to check the results:

- `WC_Tracker::get_cart_checkout_info();` 

This should show the settings you've customised in `checkout_block_attributes` and `cart_block_attributes`, for example:

```
wp> WC_Tracker::get_cart_checkout_info()
=> array(6) {
  ["cart_page_contains_cart_block"]=>
  string(3) "Yes"
  ["cart_page_contains_cart_shortcode"]=>
  string(2) "No"
  ["cart_block_attributes"]=>
  array(1) {
    ["isShippingCalculatorEnabled"]=>
    bool(false)
  }
  ["checkout_page_contains_checkout_block"]=>
  string(3) "Yes"
  ["checkout_page_contains_checkout_shortcode"]=>
  string(2) "No"
  ["checkout_block_attributes"]=>
  array(5) {
    ["showCompanyField"]=>
    bool(true)
    ["requireCompanyField"]=>
    bool(true)
    ["showPhoneField"]=>
    bool(false)
    ["requirePhoneField"]=>
    bool(true)
    ["showPolicyLinks"]=>
    bool(false)
  }
}
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Adds usage data for the of cart & checkout blocks (currently in development in WooCommmerce Blocks plugin) to the WC Tracker snapshot.

haszari committed:
add customised cart & checkout block settings to wc-tracker snapshot

haszari committed:
use anonymous functions for array_filter when finding cart etc blocks

haszari committed:
remove stale/redundant comments

Aljullu reviewed (commented):
I will defer the final approval to somebody else with more experience in WooCommerce Core codebase, but I tested it and couldn't find any issue. Thanks for working on this @haszari. :+1:

peterfabian reviewed (commented):
Not directly related to this PR, but I find it a bit inconsistent/not nice that we use direct SQL text match to test for the presence of blocks and then also parse the text using `parse_blocks`. 

Otherwise looks good to me and works fine.

haszari committed:
refactor block tracker data into standard function:
- use parse_blocks to detect block instead of SQL match
- method is now generic, can get data about any block on any woo page

haszari committed:
fix logic errors in get_blocks_from_page:
- declare closure use (dependency) on $block_name from surrounding func
- re-index the filtered array so $blocks[0] returns first filtered block

haszari commented on includes/class-wc-tracker.php:
```diff
@@ -652,16 +652,89 @@ public static function post_contains_text( $post_id, $text ) {
+		// Return any customised attributes from the first block.
+		return $cart_blocks[0]['attrs'];
+	}
+
```
Hmm, once I started refactoring to use `parse_blocks` instead of SQL match, it seemed nicer to standardise this method.

ObliviousHarmony replied:
I think I prefer @haszari's solution, as it will simplify the process of adding this functionality to other blocks.

haszari commented:
@vedanshujain @peterfabian 

Ready for review again – I've reworked this quite a bit. We're now using `parse_blocks` to detect the block and get the settings out - see new `get_block_tracker_data()` method.

I also found an issue with the previous code - `array_filter` breaks the array indexes. This bug meant that the tracker info wouldn't include the attributes if the block wasn't first on the page. This is also now fixed by using `array_values`.

So please re-review. If you retest, I recommend testing with the blocks in different places on the page, i.e. not always the first block :)

Thanks!

ObliviousHarmony commented on includes/class-wc-tracker.php:
```diff
@@ -652,33 +652,102 @@ public static function post_contains_text( $post_id, $text ) {
+		$page_blocks = self::get_all_blocks_from_page( $woo_page_name );
+
+		// Get any instances of the specified block.
+		return array_values(
```
How would you feel about just making this a loop?

haszari replied:
I am open to this .. in future :)

ObliviousHarmony reviewed (changes requested):
Looks great :) Just one small nitpick!
